File: contests/IOI/2014/Official practise/ioi14-practice/tile/judge/grader.pas
program grader;

uses tile;

function max(a, b, c: longint): longint;
begin
	if b > a then a := b;
	if c > a then a := c;
	max := a;
end;

function min(a, b, c: longint): longint;
begin
	if b < a then a := b;
	if c < a then a := c;
	min := a;
end;

procedure bye;
begin
	writeln('incorrect.');
	halt;
end;

var
	n, x, y: longint;
	t: tileType;
	s, m: longint;
	f: array[0..255, 0..255] of boolean;
	i: longint;

procedure checkxy(x, y: longint);
begin
	if (x < 0) or (x >= s) or (y < 0) or (y >= s) then bye;
	if f[x, y] then bye;
	f[x, y] := true;
end;

procedure check(x1, y1, x2, y2, x3, y3: longint);
begin
	checkxy(x1, y1);
	checkxy(x2, y2);
	checkxy(x3, y3);
	if max(x1, x2, x3) <> min(x1, x2, x3) + 1 then bye;
	if max(y1, y2, y3) <> min(y1, y2, y3) + 1 then bye;
	if (x1 = x2) and (y1 = y2) then bye;
	if (x1 = x3) and (y1 = y3) then bye;
	if (x2 = x3) and (y2 = y3) then bye;
end;

begin
	read(n, x, y);
	s := 1 shl n;
	placeTile(n, x, y, t);
	m := s * s div 3;
	for i := 0 to m - 1 do
		check(t[i, 0], t[i, 1], t[i, 2], t[i, 3], t[i, 4], t[i, 5]);
	writeln(y, '~', n, '~', x, '~', n * x * y);

end.

File: contests/IOI/2014/Official practise/ioi14-practice/tile/attachment/grader.cpp
#ifdef NDEBUG
#undef NDEBUG
#endif
#include<cassert>
#include<cstdio>
#include<cstdlib>
#include"tile.h"

static inline int max(int a,int b,int c)
{
	if(b>a)a=b;
	if(c>a)a=c;
	return a;
}

static inline int min(int a,int b,int c)
{
	if(b<a)a=b;
	if(c<a)a=c;
	return a;
}

static void bye()
{
	puts("incorrect.");
	exit(0);
}

static int s;

static inline void checkxy(int x,int y)
{
	static char f[256][256];
	if(x<0||x>=s||y<0||y>=s||f[x][y]!=0)bye();
	f[x][y]=1;
}

static void check(int x1,int y1,int x2,int y2,int x3,int y3)
{
	checkxy(x1,y1);
	checkxy(x2,y2);
	checkxy(x3,y3);
	if(max(x1,x2,x3)!=min(x1,x2,x3)+1)bye();
	if(max(y1,y2,y3)!=min(y1,y2,y3)+1)bye();
	if(x1==x2&&y1==y2)bye();
	if(x1==x3&&y1==y3)bye();
	if(x2==x3&&y2==y3)bye();
}

int main()
{
	static int t[21845][6];
	int n,x,y;
	assert(scanf("%d%d%d",&n,&x,&y)==3);
	assert(1<=n&&n<=8);
	s=1<<n;
	assert(0<=x&&x<s);
	assert(0<=y&&y<s);
	placeTile(n,x,y,t);
	const int m=s*s/3;
	for(int i=0;i<m;i++)check(t[i][0],t[i][1],t[i][2],t[i][3],t[i][4],t[i][5]);
	puts("correct.");
	return 0;
}
